### PR TITLE
feat: Add optional XAML Trimming

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Feature

## What is the new behavior?

Adds a XAML trimming feature which removes unused XAML resources, based on what the application currently uses.

As of Uno 3.9, for a `dotnet new unoapp` created app:

|                      | w/o XAML Trimming | w/ XAML Trimming |
| -------------------- | ----------------- | --------------- |
| Total IL Payload     |            12.9MB |         9.12 MB |
| dotnet.wasm          |              53MB |       28.9MB MB |

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: doc/articles/features/resources-trimming.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# XAML Resource Trimming
+
+XAML Resource and Binding trimming is a optional feature used to reduce the size of the final payload of an Uno Platform application.
```
Comment (davidjohnoliver) [inline doc/articles/features/resources-trimming.md:3]:
```suggestion
XAML Resource and Binding trimming is an optional feature used to reduce the size of the final payload of an Uno Platform application.
```

File: doc/articles/features/resources-trimming.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+In order for an application to enable resources trimming, the following needs to be added in the project file:
+
+```xml
+<PropertyGroup>
+    <UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>
+</PropertyGroup>
+```
+
+## Enabling XAML Resources trimming for libraries
+For libraries to be elligible for resources trimming, the `UnoXamlResourcesTrimming` must also be added.
```
Comment (davidjohnoliver) [inline doc/articles/features/resources-trimming.md:20]:
```suggestion
For libraries to be eligible for resources trimming, the `UnoXamlResourcesTrimming` tag must also be added.
```

File: doc/articles/features/resources-trimming.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+For instance, if your application relies on the `XamlReader` class, trimmed controls will not be available and will fail to load.
+
+If XAML trimming is still needed, the [IL Linker configuration](using-il-linker-webassembly.md) can be adjusted to keep controls individually or by namespace.
+
+### Size is not reduced even if enabled
+The IL Linker tool is used to implement this feature, and can be [controlled with its configuration file](using-il-linker-webassembly.md).
+
+For instance, if the linker configuration file contains `<assembly fullname="uno.ui" />`, none of the UI Controls will be excluded, and the final app size will remain close as without trimming.
+
+Note that for libraries, Uno 3.9 or later must be used to build the library, as additional metadata needs to be added a compile time. 3.8 and earlier libraries can be used in any case, but won't be elligible for trimming and may degrade the trimming phase effect.
```
Comment (davidjohnoliver) [inline doc/articles/features/resources-trimming.md:36]:
```suggestion
Note that for libraries, Uno 3.9 or later must be used to build the library, as additional metadata needs to be added at compile time. 3.8 and earlier libraries can be used in any case, but won't be eligible for trimming and may degrade the trimming phase effect.
```

File: doc/articles/toc.yml [hunk 4]
```diff
@@ -313,6 +313,8 @@
           href: uno-fluent-assets.md
         - name: Lottie animations
           href: features/Lottie.md
+        - name: Working XAML Trimming
```
Comment (davidjohnoliver) [inline doc/articles/toc.yml:316]:
```suggestion
        - name: Working with XAML Trimming
```

File: doc/articles/uno-development/Uno-UI-XAML-ResourceTrimming.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+- During the source generation, the tooling generates a `LinkerHints` class, which contains a set of properties for all `DependencyObject` inheriting classes.
+- The source generation creates XAML Resources and Bindable Metadata code that conditionally uses those classes behind `LinkerHints` properties. 
+- The tooling also embeds an ILLinker substitution file allowing the linker to unconditionally remove the code that conditionally references those properties. For instance, for `LinkerHints.Is_Windows_UI_Xaml_Controls_Border_Available`, any block of `if (LinkerHints.Is_Windows_UI_Xaml_Controls_Border_Available)` will be removed when the `--feature Is_Windows_UI_Xaml_Controls_Border_Available false` parameter is provided to the linker. 
+
+Then the multiple passes of IL Linker are done:
+- The first pass runs the IL Linker with all XAML resources and Binding Metadata disabled by setting all `LinkerHints` properties to false. This removes all code directly associated to those Bindable Medata and XAML Resources. This has the effect of only keeping framework code which is directly referenced from user code.
```
Comment (davidjohnoliver) [inline doc/articles/uno-development/Uno-UI-XAML-ResourceTrimming.md:20]:
```suggestion
- The first pass runs the IL Linker with all XAML resources and Binding Metadata disabled by setting all `LinkerHints` properties to false. This removes all code directly associated to those Bindable Metadata and XAML Resources. This has the effect of only keeping framework code which is directly referenced from user code.
```

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 6]
```diff
@@ -83,8 +83,15 @@ internal object Lookup(string key)
 		/// and can be removed as breaking change later.</remarks>
 		public bool Insert(object key, object value)
 		{
-			Set(new ResourceKey(key), value, throwIfPresent: false);
-			return true;
+			if (key is { })
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:86]: Does this match UWP behavior?

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:86]: It's not the same behavior, it's a temporary behavior to allow for the xaml trimming generated code to still work and not be too complex. I'll mention it in an issue.

**Comment by jeromelaban** [comment]
@merifyion backport release/stable/3.9

**Comment by jeromelaban** [comment]
@mergifyio backport release/stable/3.9

**Pull request merged**